### PR TITLE
Add adjustable password length and tabu list for recently used passwords

### PR DESCRIPTION
**Pull request by seitenbau-govdata**
We made several improvements to this plugin that are based on each other. We have included all improvements in this PR, it's a bit messy, but if needed, we can talk and maybe find a good way to split them up.

What we have done:

- added tests
- added translations (and included a german translation)
- add config declaration
- bugfix for newer CKAN, session object changed and have no "invalidate" method anymore
- if used with https://github.com/keitaroinc/ckanext-saml2auth, we patch the other extension to use the password generation of this extension - so generated passwords comply with the requirements defined here

### CODE REVIEW

File: ckanext/security/constants.py [hunk 1]
```diff
@@ -0,0 +1 @@
+PLUGIN_EXTRAS_TABULIST_KEY = 'password_tabulist'
```
Comment (ThrawnCA) [inline ckanext/security/constants.py:1]: This reads more clearly if "tabu" and "list" are separated into two words, so that it doesn't resemble table-related words like "tabular". Or change to "blacklist".

File: ckanext/security/helpers.py [hunk 2]
```diff
@@ -1,5 +1,40 @@
+from ckan.common import _
```
Comment (ThrawnCA) [inline ckanext/security/helpers.py:1]: This should come from `ckan.plugins.toolkit` rather than `ckan.common`

File: ckanext/security/helpers.py [hunk 3]
```diff
@@ -1,5 +1,40 @@
...
 from ckan.plugins.toolkit import asbool, config
+import string, random
```
Comment (ThrawnCA) [inline ckanext/security/helpers.py:3]: The `random` library is not cryptographically strong. Consider the `secrets` library instead.

File: ckanext/security/plugin/i18n/ckanext-security.pot [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+"Content-Type: text/plain; charset=utf-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+"Generated-By: Babel 2.15.0\n"
+
+#: ckanext/security/authenticator.py:181
+msgid "Login failed. Bad username or password."
+msgstr ""
+
+#: ckanext/security/helpers.py:6
+msgid " Your password must not be the same as any of your last {} passwords."
```
Comment (ThrawnCA) [inline ckanext/security/plugin/i18n/ckanext-security.pot:26]: Why does this start with a space?

File: ckanext/security/plugin/i18n/ckanext-security.pot [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+msgid "No user specified"
+msgstr ""
+
+#: ckanext/security/utils.py:51
+msgid "Not authorized to see this page"
+msgstr ""
+
+#: ckanext/security/utils.py:207
+msgid ""
+"That's a valid code. Your authenticator\n"
```
Comment (ThrawnCA) [inline ckanext/security/plugin/i18n/ckanext-security.pot:39]: Having a line break in the message ID is inelegant.

File: ckanext/security/templates/user/edit_user_form.html [hunk 6]
```diff
@@ -16,3 +16,23 @@
 {% endif %}
 
 {% endblock %}
+
+{% block change_password %}
```
Comment (ThrawnCA) [inline ckanext/security/templates/user/edit_user_form.html:20]: It's not immediately obvious why this block is being overridden, because at first glance it's just re-adding the old and new password fields. Can a comment be added about what is changing?

File: ckanext/security/tests/test_action.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+        user_obj.password = "new_password3"
+        action._append_password(context, user_obj) # will replace the last entry in tabu list
+
+        # THEN
+        actual_user = tk.get_action('user_show')(context, {'id': user['name'], 'include_plugin_extras': True})
+        assert 'plugin_extras' in actual_user
+        assert PLUGIN_EXTRAS_TABULIST_KEY in actual_user['plugin_extras']
+        assert len(actual_user['plugin_extras'][PLUGIN_EXTRAS_TABULIST_KEY]) == 2
+        assert pbkdf2_sha512.verify("new_password3", actual_user['plugin_extras'][PLUGIN_EXTRAS_TABULIST_KEY][0])
```
Comment (ThrawnCA) [inline ckanext/security/tests/test_action.py:39]: Ideally the test would actually verify which passwords result in errors.

File: ckanext/security/tests/test_validators.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        """Check if the password length is validated correctly"""
+        # WHEN + THEN
+        validators.user_password_validator("pw", {"pw": "Aa_123456789"}, None, None)
+
+    @pytest.mark.ckan_config(u'ckanext.security.min_password_length', u'12')
+    def test_invalid_password_length_is_too_short(self):
+        """Check if the password length is validated correctly"""
+        # WHEN + THEN
+        with pytest.raises(Invalid):
+            validators.user_password_validator("pw", {"pw": "Aa_123"}, None, None)
```
Comment (ThrawnCA) [inline ckanext/security/tests/test_validators.py:40]: This should test a range of cases, especially edge cases. Eg if the minimum is 12, we should verify that a 12-character password is okay and 11 is not. Consider using a parameterised test.

File: ckanext/security/tests/test_validators.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    # Check if one missing character class still validates.
+    def test_invalid_no_uppercase_letters(self):
+        # WHEN + THEN
+        validators.user_password_validator("pw", {"pw": "aa_123456789"}, None, None)
```
Comment (ThrawnCA) [inline ckanext/security/tests/test_validators.py:45]: Test name is wrong since this is actually valid.

File: ckanext/security/validators.py [hunk 10]
```diff
@@ -1,17 +1,32 @@
...
 from ckan.lib.navl.dictization_functions import Missing, Invalid
+from ckan.plugins import toolkit as tk
+from ckanext.security.constants import PLUGIN_EXTRAS_TABULIST_KEY
+from passlib.hash import pbkdf2_sha512
+
+DEFAULT_MIN_PASSWORD_LENGTH = 10
+
+PASSWORD_ERROR = ("Your password must be {} characters or longer, and consist of at least three" +
+                  " of the following four character sets: uppercase characters, lowercase characters," +
+                  " digits, punctuation & special characters.")
```
Comment (ThrawnCA) [inline ckanext/security/validators.py:16]: There should be some explanation of which special characters are accepted.

File: ckanext/security/plugin/i18n/ckanext-security.pot (hunk 5 above)
Comment (seitenbau-govdata) [reply ckanext/security/plugin/i18n/ckanext-security.pot:39]: Absolutely, fixed it. We just ran the translation tool and the untranslated strings already had line breaks in them.

File: ckanext/security/tests/test_validators.py (hunk 9 above)
Comment (seitenbau-govdata) [reply ckanext/security/tests/test_validators.py:45]: oh yes, copy & paste typo. Thank you!

**Comment by seitenbau-govdata** [comment]
@ThrawnCA Thank you very much for your review!

File: ckanext/security/validators.py [hunk 11]
```diff
@@ -1,17 +1,32 @@
 # encoding: utf-8
-import six
 import string
 
+import six
```
Comment (ThrawnCA) [inline ckanext/security/validators.py:4]: `six` is a good library, but...is this plugin actually compatible with Python 2? If not, then there's no point.

Comment (ThrawnCA) [reply ckanext/security/validators.py:4]: > git blame says you introduced it 2021

Sure. In 2021, CKAN 2.8, and thus Python 2, were still current.

Comment (seitenbau-govdata) [reply ckanext/security/validators.py:4]: I just wondered where six came from, that's why I mentioned it. :) Please don't take that as criticism.
I just open an issue to track this so it can be done in another PR.
